Remotion/Core/UnitTests/Context/AsyncLocalStorageProviderTest.cs

```diff
@@ -102,42 +103,37 @@ public void SetData_FromOtherThread_DoesNotOverride ()
     public void SetData_OnThreadPoolThread_IsResetForNextWorkItem ()
     {
       object result = string.Empty;
-      int threadId1 = -1;
-      int threadId2 = -2;
-      ThreadPool.GetMinThreads (out var minWorkerThreads, out var minCompletionPortThreads);
-      ThreadPool.GetMaxThreads (out var maxWorkerThreads, out var maxCompletionPortThreads);
-      ThreadPool.GetAvailableThreads (out var availableWorkerThreads, out _);
-      try
-      {
-        ThreadPool.SetMinThreads (1, 0);
-        ThreadPool.SetMaxThreads (maxWorkerThreads - availableWorkerThreads + 1, 0);
-        using var resetEvent = new AutoResetEvent (false);
 
+      using var resetEvent = new AutoResetEvent (false);
+
+      var visitedThreadPoolThreads = new HashSet<int>();
+      var completed = false;
+
+      while (!completed)
+      {
         ThreadPool.QueueUserWorkItem (
             _ =>
             {
-              threadId1 = Thread.CurrentThread.ManagedThreadId;
-              _provider.SetData (c_testKey, "1");
-              resetEvent.Set();
-            });
-        resetEvent.WaitOne();
-        ThreadPool.QueueUserWorkItem (
-            _ =>
-            {
-              threadId2 = Thread.CurrentThread.ManagedThreadId;
-              result = _provider.GetData (c_testKey);
+              var threadPoolThreadId = Thread.CurrentThread.ManagedThreadId;
+              var didNotVisitThreadPoolThreadYet = visitedThreadPoolThreads.Add (threadPoolThreadId);
+              
+              if (didNotVisitThreadPoolThreadYet)
+              {
+                _provider.SetData (c_testKey, "1");
+              }
+              else
+              {
+                result = _provider.GetData (c_testKey);
+                completed = true;
+              }
+
+              Thread.MemoryBarrier();
               resetEvent.Set();
             });
         resetEvent.WaitOne();
```

Comment:
Say, should we add a timeout to the WaitOne and fail the test if it's reached so the code can't hang the build if there's a thread-starvation bug or just environment condition? And if if so, new ticket (please create and schedule) or right here?

Reply:
I don't think that such a case will ever come up but I've added a timeout of 3sec onto the wait.